dvc/remote/local/__init__.py

```diff
@@ -566,3 +591,69 @@ def unprotect(path_info):
     @staticmethod
     def protect(path_info):
         os.chmod(path_info.path, stat.S_IREAD | stat.S_IRGRP | stat.S_IROTH)
+
+    @classmethod
+    def _append_unpacked_suffix(cls, string):
+        return string + cls.UNPACKED_DIR_SUFFIX
+
+    def _get_unpacked_dir_info(self, checksum):
+        unpacked_dir_info = self.checksum_to_path_info(checksum)
+        unpacked_dir_info.path = self._append_unpacked_suffix(
+            unpacked_dir_info.path
+        )
+
+        return unpacked_dir_info
+
+    def _update_unpacked_dir(self, checksum, dir_info):
+        unpacked_dir_info = self._get_unpacked_dir_info(checksum)
+
+        exists = self.exists(unpacked_dir_info)
+        if exists:
+            if self.state.get(unpacked_dir_info):
+                return
+
+            shutil.rmtree(unpacked_dir_info.path)
+
+        self._create_unpacked_dir(checksum, dir_info, unpacked_dir_info)
+
+    def _create_unpacked_dir(self, checksum, dir_info, unpacked_dir_info):
+        self.makedirs(unpacked_dir_info)
+        for entry in dir_info:
+            entry_cache_info = self.checksum_to_path_info(
+                entry[self.PARAM_CHECKSUM]
+            )
+            relpath = entry[self.PARAM_RELPATH]
+
+            unpacked_entry_info = copy(entry_cache_info)
+            unpacked_entry_info.path = self.ospath.join(
+                unpacked_dir_info.path, relpath
+            )
+            self.link(entry_cache_info, unpacked_entry_info, "hardlink")
```

Comment:
Let's say we are trying to create unpacked dir and we find that some cache file doesn't exist. Then we `link` will raise an exception, right? And it will simply break dvc and prevent it from doing anything, right? If so, it feels like we should not break on that, but rather simply not create this dir and fall back to manual check,

Reply:
Agree, thought i think this problem will be detected earlier, when creating cache file. 
Situation that you describe is still valid when os does not support hardlink.

Reply:
Good point! But I'm talking about `dvc fetch` that failed to download some cache files and then you `dvc checkout`. Currently that will be detected while manually checking cache files, but with your patch it will simply fail on `link` here.

Reply:
Fixed by catching exception on unpacked dir creation.

Reply:
Fixed by catching exception on unpacked dir creation